engine/src/main/java/org/pentaho/di/trans/steps/transexecutor/TransExecutor.java

```diff
@@ -272,31 +285,98 @@ Trans createInternalTrans() throws KettleException {
+    //The values of the incoming fields from the previous step.
+    if ( incomingFieldValues == null ) {
+      incomingFieldValues = new ArrayList<String>();
+    }
+
     // Set parameters, when fields are used take the first row in the set.
     TransExecutorParameters parameters = meta.getParameters();
 
-    Trans internalTrans = getData().getExecutorTrans();
+    // A map where the final parameters and values are stored.
+    Map<String, String> resolvingValuesMap = new HashMap<String, String>();
     for ( int i = 0; i < parameters.getVariable().length; i++ ) {
-      String fieldName = parameters.getField()[ i ];
-      String inputValue = parameters.getInput()[ i ];
-      String value;
-      // Take the value from an input row or from a static value?
-      if ( fieldName != null && !"".equals( fieldName ) ) {
-        int idx = getInputRowMeta().indexOfValue( fieldName );
-        if ( idx < 0 ) {
-          throw new KettleException( BaseMessages.getString(
-            PKG, "TransExecutor.Exception.UnableToFindField", fieldName ) );
+      resolvingValuesMap.put( parameters.getVariable()[i], null );
+    }
+
+    //The names of the "Fields to use".
+    List<String> fieldsToUse = new ArrayList<String>();
+    if ( parameters.getField() != null ) {
+      fieldsToUse = Arrays.asList( parameters.getField() );
+    }
+
+    //The names of the incoming fields from the previous step.
+    List<String> incomingFields = new ArrayList<String>();
+    if ( data.getInputRowMeta() != null ) {
+      incomingFields = Arrays.asList( data.getInputRowMeta().getFieldNames() );
+    }
+
+    //The values of the "Static input value".
+    List<String> staticInputs = Arrays.asList( parameters.getInput() );
+
+    /////////////////////////////////////////////
+    //If "Fields to use" ARE provided.
+    for ( int i = 0; i < fieldsToUse.size(); i++ ) {
+      try {
+        String value = null;
+        if ( incomingFields.contains( fieldsToUse.get( i ) )
+            && ( value = incomingFieldValues.get( incomingFields.indexOf( fieldsToUse.get( i ) ) ) ) != null ) {
+          //Set the value to the first parameter in the resolvingValuesMap.
+          resolvingValuesMap.put( (String) resolvingValuesMap.keySet().toArray()[i], value );
+        } else {
+          //Set the value to the first parameter in the resolvingValuesMap.
+          resolvingValuesMap.put( (String) resolvingValuesMap.keySet().toArray()[i], staticInputs.get( i ) );
+        }
+      } catch ( Exception e ) {
+        //Set the value to the first parameter in the resolvingValuesMap.
+        resolvingValuesMap.put( (String) resolvingValuesMap.keySet().toArray()[i], null );
+      }
+    }
+    /////////////////////////////////////////////
+
+    /////////////////////////////////////////////
+    //If "Fields to use" ARE NOT provided and "Inherit all variables from transformation" IS checked.
```

Comment:
It looks like this is all or nothing. If you have one "field to use", then you look to resolve all parameters for the input row.   The way we did it on AEL we would first look to match the parameter from the "field to use" and if we could not find the parameter there we would then check for the parameter being defined on the transform. If the inherent variables where set.